### PR TITLE
Add documentation of security implications

### PR DESCRIPTION
**Pull request by srittau**
The maintainers section is still missing, but the remaining sections are up for scrutiny.

Cc @AlexWaygood @hauntsaninja @Avasam

### CODE REVIEW

File: SECURITY.md
Comment (JelleZijlstra) [inline SECURITY.md:15]:
```suggestion
To ensure that a compromised typeshed repository can't lead to compromised
```

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, great idea to add this!

Another area to cover could be the security of the upload process itself: what are we doing to prevent an issue where a malicious party could gain the ability to upload stub packages using the uploader's identity? (Relevant real-world example: https://blog.pypi.org/posts/2024-07-08-incident-report-leaked-admin-personal-access-token/.)

**Comment by srittau** [comment]
Ironically, I forgot to ping @Akuli, although it was #138 that prompted me to write this. Sorry!

@JelleZijlstra What can we do about that? In a related matter: I can't check, but do we require 2FA to be enabled for this org/repository? If not, we should definitely enable it.

**Comment by ilevkivskyi** [comment]
> but do we require 2FA to be enabled for this org/repository?

I am 90% sure I enabled it a while ago, I will double-check now.

**Comment by JelleZijlstra** [comment]
> What can we do about that?

Haven't thought too much about it, but it looks like stub packages are owned only by typeshed-bot (e.g., https://pypi.org/project/types-requests/). That's good because it means that if someone gains access to my PyPI account, they can't use it to upload a malicious stubs package. However, we should also review what kind of authentication typeshed-bot uses to upload packages.

**Comment by ilevkivskyi** [comment]
> I am 90% sure I enabled it a while ago, I will double-check now.

Actually it turns out "Require two-factor authentication for everyone in the typeshed-internal organization" was off. I just turned it on. (I also checked that all four members have 2FA enabled ;-))

**Comment by ilevkivskyi** [comment]
> what kind of authentication typeshed-bot uses to upload packages

It uses a PyPI token. FWIW we can rotate it from time to time. (There is even a GHA to test that PyPI token works.)

**Comment by srittau** [comment]
We should probably document the typeshed-bot user in this file as well. Commits to the PR or text suggestions in the comments welcome. Otherwise I will look into this tomorrow.

File: SECURITY.md
Comment (Akuli) [inline SECURITY.md:34]: It would be nice to expand upon this and actually describe and explain the possible attack in https://github.com/typeshed-internal/stub_uploader/pull/61#discussion_r979327370. This kind of hints at how the attack goes, but IMO does not describe it in enough detail. It is the biggest security issue in stub_uploader I'm aware of, and typeshed maintainers need to be aware of it.

**Comment by Avasam** [comment]
I got nothing to add. You covered everything I was already aware of 👍

File: SECURITY.md
Comment (srittau) [reply SECURITY.md:34]: I'd like to hold off documenting this here at the moment. I kept this section quite vague for the moment, mostly because I think we should resolve and implement #90 first and can then discuss the measures and reasons in this document.

**Comment by srittau** [comment]
I've added the maintainers section, this is ready for a full review now. As mentioned above, I'd like to punt @Akuli's suggestion to describe the attack to later when we've decided what to do in the future in #90.

File: SECURITY.md
Comment (JelleZijlstra) [inline SECURITY.md:1]: A SECURITY.md should by convention include a way to privately report security vulnerabilities. For typing-extensions (https://github.com/python/typing_extensions/blob/main/SECURITY.md), I redirected reports to GitHub's private reporting feature, but for this repo I don't have permissions to turn on that feature.

So I'd suggest:

- Someone with the right permissions (@ilevkivskyi?) should go to https://github.com/typeshed-internal/stub_uploader/security and turn on "Private vulnerability reporting"
- In this PR we should add a line saying that any vulnerabilities should be reported to https://github.com/typeshed-internal/stub_uploader/security

We should also add a SECURITY.md to the typeshed repo pointing here.

File: SECURITY.md
Comment (JelleZijlstra) [inline SECURITY.md:17]:
```suggestion
Typeshed maintainers without commit rights are still considered maintainers
```

File: SECURITY.md
Comment (ilevkivskyi) [reply SECURITY.md:1]: OK, I can do this later today, or @JukkaL should be able to do this.

Comment (srittau) [reply SECURITY.md:1]: I've added such a section to the document for now, mostly taking the wording from typing-extension.

Comment (ilevkivskyi) [reply SECURITY.md:1]: OK, I enabled "Private vulnerability reporting" for this repo.

**Comment by srittau** [comment]
Are there any more notes?

**Review by JelleZijlstra — APPROVED** [review]

File: SECURITY.md
Comment (AlexWaygood) [inline SECURITY.md:32]: "Possible attack vector" is a singular noun

```suggestion
Another possible attack vector is dependencies of stub packages.
```

File: SECURITY.md
Comment (AlexWaygood) [inline SECURITY.md:49]:
```suggestion
upload new versions of the packages, in case the account of a
maintainer becomes compromised.
```

**Review by AlexWaygood — APPROVED** [review]

File: SECURITY.md
Comment (Akuli) [inline SECURITY.md:50]: Couldn't a maintainer account simply edit CI config to store the token somewhere? Or is this about maintainers who do not have commit access, like me?

File: SECURITY.md
Comment (Akuli) [reply SECURITY.md:42]: Maybe:

```suggestion
  dependencies are vetted to be from a trusted source.

Having one global allowlist is not ideal, because it allows somewhat obscure
dependencies to be added to popular stub packages whose security matters
the most. This may be changed in the near future.
```

File: SECURITY.md
Comment (JelleZijlstra) [reply SECURITY.md:50]: I think it's about the PyPI accounts, not GitHub accounts. If a GitHub account with commit access (e.g., mine) gets compromised, it could obviously be used to upload a malicious package just by changing the code. But if someone compromises my PyPI account, they can't use it to upload malicious stub packages. (They could use it to upload a malicious version of typing-extensions, which is probably worse, but that's a different repo than this one.)

File: SECURITY.md
Comment (Akuli) [inline SECURITY.md:49]:
```suggestion
The packages are owned solely by that user, so that no other PyPI user can
upload new versions of the packages, in case the PyPI account of a
```

**Review by Akuli — APPROVED** [review]

**Pull request merged**
